tarantool/msgpack_ext/types/datetime.py

```diff
@@ -0,0 +1,247 @@
+# - Etc/GMT+2
+# - Etc/GMT+3
+# - Etc/GMT+4
+# - Etc/GMT+5
+# - Etc/GMT+6
+# - Etc/GMT+7
+# - Etc/GMT+8
+# - Etc/GMT+9
+# - Etc/GMT-1
+# - Etc/GMT-10
+# - Etc/GMT-11
+# - Etc/GMT-12
+# - Etc/GMT-13
+# - Etc/GMT-14
+# - Etc/GMT-2
+# - Etc/GMT-3
+# - Etc/GMT-4
+# - Etc/GMT-5
+# - Etc/GMT-6
+# - Etc/GMT-7
+# - Etc/GMT-8
+# - Etc/GMT-9
+# - Europe/Kyiv
+# - MET
+# - MST7MDT
+# - PST8PDT
+#
+# They are transformed to tzoffset based on pytz info.
+    tzoffset = compute_offset(timestamp)
+
+    if (timestamp.tz is None) or (timestamp.tz.zone is None):
+        tzindex = 0
+    else:
+        if timestamp.tz.zone in tt_timezones.timezoneToIndex:   
+            tzindex = tt_timezones.timezoneToIndex[timestamp.tz.zone]   
+        else:  
+            warn(f'pytz timezone {timestamp.tz} is not supported by Tarantool, '
+                 f'using tzoffset={tzoffset} instead', MsgpackWarning)
+
+            tzindex = 0
+
+    return tzoffset, tzindex
+
+def is_ambiguous_tz(tt_tzinfo):
+    return (tt_tzinfo['category'] & tt_timezones.TZ_AMBIGUOUS) != 0
+
+def tarantool_to_pytz(tzindex, error_class=ValueError):
+# https://raw.githubusercontent.com/tarantool/tarantool/9ee45289e01232b8df1413efea11db170ae3b3b4/src/lib/tzcode/timezones.h
+#
+# There are several possible timezone types in Tarantool.
+# Abbreviated timezones are a bit tricky since they could be ambiguous.
+# Tarantool itself do not support creating datetime with ambiguous timezones:
+#
+# Tarantool 2.10.1-0-g482d91c66
+# 
+# tarantool> datetime.new({tz = 'BST'})
+# ---
+# - error: 'builtin/datetime.lua:477: could not parse ''BST'' - ambiguous timezone'
+# ...
+#
+# pytz version 2022.2.1 do not support most of Tarantool abbreviated timezones
+# (except for CET, EET EST, GMT, HST, MST, UTC, WET). Since Tarantool sources
+# provide offset info for abbreviated timezones, we use pytz.FixedOffset instead.
+#
+# https://stackoverflow.com/questions/30315485/pytz-return-olson-timezone-name-from-only-a-gmt-offset
+
+    if tzindex not in tt_timezones.indexToTimezone:
+        raise error_class(f'Unknown tzindex {tzindex}')
+    tzname = tt_timezones.indexToTimezone[tzindex]
+
+    try:
+        tzinfo = pytz.timezone(tzname)
+    except pytz.exceptions.UnknownTimeZoneError:
+        tt_tzinfo = tt_timezones.timezoneAbbrevInfo[tzname]
+
+        if is_ambiguous_tz(tt_tzinfo):
+            raise error_class(f'Failed to create datetime with ambiguous timezone {tzname}')
+
+        tzinfo = pytz.FixedOffset(tt_tzinfo['offset'])
+
+    return tzinfo
+
+def msgpack_decode(data):
+    cursor = 0
+    seconds, cursor = get_bytes_as_int(data, cursor, SECONDS_SIZE_BYTES)
+
+    if len(data) > SECONDS_SIZE_BYTES:
+        nsec, cursor     = get_bytes_as_int(data, cursor, NSEC_SIZE_BYTES)
+        tzoffset, cursor = get_bytes_as_int(data, cursor, TZOFFSET_SIZE_BYTES)
+        tzindex, cursor  = get_bytes_as_int(data, cursor, TZINDEX_SIZE_BYTES)
+    else:
+        nsec = 0
+        tzoffset = 0
+        tzindex = 0
```

Comment:
This will be less error-prone:
```suggestion
    if len(data) == SECONDS_SIZE_BYTES + NSEC_SIZE_BYTES + TZOFFSET_SIZE_BYTES + TZINDEX_SIZE_BYTES:
        nsec, cursor     = get_bytes_as_int(data, cursor, NSEC_SIZE_BYTES)
        tzoffset, cursor = get_bytes_as_int(data, cursor, TZOFFSET_SIZE_BYTES)
        tzindex, cursor  = get_bytes_as_int(data, cursor, TZINDEX_SIZE_BYTES)
    else if len(data) == SECONDS_SIZE_BYTES:
        nsec = 0
        tzoffset = 0
        tzindex = 0
    else:
         raise errors_class(f"Failed to decode bla-bla-bla, invalid length %d", len(data))
```

Reply:
Reworked

Reply:
Why not make an explicit equal, insteat of greater?
`len(data) == SECONDS_SIZE_BYTES + NSEC_SIZE_BYTES + TZOFFSET_SIZE_BYTES + TZINDEX_SIZE_BYTES`

Reply:
Thank you, used this approach.